src/testnet/helium/event_dispatcher.rs

```diff
@@ -0,0 +1,242 @@
+    any_event_observers_lookup: HashSet<u16>,
+}
+
+impl EventDispatcher {
+
+    pub fn new() -> EventDispatcher {
+        EventDispatcher {
+            registered_observers: vec![],
+            contract_events_observers_lookup: HashMap::new(),
+            assets_observers_lookup: HashMap::new(),
+            stx_observers_lookup: HashSet::new(),
+            any_event_observers_lookup: HashSet::new(),
+        }
+    }
+
+    pub fn process_receipts(&mut self, receipts: &Vec<StacksTransactionReceipt>, chain_tip: &StacksBlock, chain_tip_info: &StacksHeaderInfo) {
+        let mut dispatch_matrix: Vec<HashSet<usize>> = self.registered_observers.iter().map(|_| HashSet::new()).collect();
+        let mut events: Vec<(Txid, &StacksTransactionEvent)> = vec![];
+        let mut i: usize = 0;
+        for artifact in receipts.iter() {
+            let tx_hash = artifact.transaction.txid();
+            for event in artifact.events.iter() {
+                match event {
+                    StacksTransactionEvent::SmartContractEvent(event_data) => {
+                        if let Some(observer_indexes) = self.contract_events_observers_lookup.get(&event_data.key) {
+                            for o_i in observer_indexes {
+                                dispatch_matrix[*o_i as usize].insert(i);
+                            }
+                        }
+                    },
+                    StacksTransactionEvent::STXEvent(STXEventType::STXTransferEvent(_)) |
+                    StacksTransactionEvent::STXEvent(STXEventType::STXMintEvent(_)) |
+                    StacksTransactionEvent::STXEvent(STXEventType::STXBurnEvent(_)) => {
+                        for o_i in &self.stx_observers_lookup {
+                            dispatch_matrix[*o_i as usize].insert(i);
+                        }
+                    },
+                    StacksTransactionEvent::NFTEvent(NFTEventType::NFTTransferEvent(event_data)) => {
+                        self.update_dispatch_matrix_if_observer_subscribed(&event_data.asset_identifier, i, &mut dispatch_matrix);
+                    },
+                    StacksTransactionEvent::NFTEvent(NFTEventType::NFTMintEvent(event_data)) => {
+                        self.update_dispatch_matrix_if_observer_subscribed(&event_data.asset_identifier, i, &mut dispatch_matrix);
+                    },
+                    StacksTransactionEvent::FTEvent(FTEventType::FTTransferEvent(event_data)) => {
+                        self.update_dispatch_matrix_if_observer_subscribed(&event_data.asset_identifier, i, &mut dispatch_matrix);
+                    },
+                    StacksTransactionEvent::FTEvent(FTEventType::FTMintEvent(event_data)) => {
+                        self.update_dispatch_matrix_if_observer_subscribed(&event_data.asset_identifier, i, &mut dispatch_matrix);
+                    },
+                }
+                events.push((tx_hash, event));
+                i += 1;
```

Comment:
I think this index increment needs to be placed beneath the `any_event_observers_lookup` loop, otherwise getting out of bounds panics.

Reply:
Addressed with latest